Access for unenrolled learners.

symbolist opened the pull request:
https://github.com/edx/edx-platform/pull/18720 was reverted because of a bug. This PR adds back the changes with the bug fixed.

**Description**:

Adds CouseModule.course_visibility and XBlock.public_view() for unenrolled users access to courses.

The course_visiblity field can have one of three values:
1. private (default): This keeps the standard access rules.
2. public_outline: Allows unenrolled and anonymous users access to the outline.
3. public: Allows unenrolled and anonymous users access to both outline and
   course content.

When an unenrolled user accesses course content, instead of student_view(),
public_view() is used. A default implementation is provided for XBlocks
which do not implement this view. The public_view() must not have any
functionality which assumes the presence of a valid User and should show
a readonly only interface for the XBlock content.

Description goes here. e.g. This PR contains the LibraryContent XBlock, which allows to display library content in a course.

**JIRA tickets**: 

https://openedx.atlassian.net/browse/EDUCATOR-3705
https://openedx.atlassian.net/browse/EDUCATOR-3706

**Discussions**: 

Proposal: https://github.com/edx/edx-platform/pull/18134
PR that was reverted: https://github.com/edx/edx-platform/pull/18720.

**Sandbox URL**: https://pr19284.sandbox.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/course/

edx-webhook commented:
Thanks for the pull request, @symbolist! I've created [OSPR-2814](https://openedx.atlassian.net/browse/OSPR-2814) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@symbolist Thank you for your contribution. Please let me know once this can be looked at.

symbolist commented:
@ormsbee @pomegranited See the second commit for the fix.

While working on this I realized that the current behavior is that while adding a team member to a course does enroll them in, it is still possible for unenrolled team members to access the courseware. From the RCA conversation, my understanding was that the later was only allowed for global staff (where User.is_staff = True). So just to make sure we are covering all cases, I have refactored the course home and courseware index access check tests and added all the possible combinations I could think of. Are there any which are missing or do not meet expectations?

I have added a global staff (email: edx@example.com) and an unenrolled staff for the demo course (email: unenrolled_staff@example.com).

symbolist commented:
Hi @natabene, thanks for checking in! @ormsbee reviewed https://github.com/edx/edx-platform/pull/18720 which was later reverted so my understanding is that he will be reviewing this one as well.

natabene commented:
@symbolist  Got it, please ping him or me once it is ready for engineering review by edX.

symbolist commented:
jenkins run bokchoy

symbolist commented:
@ormsbee This is ready for an engineering review from edX. Just to be on the safe side I have also changed the enrollment check in the milestones transformer.

I do have one question. There are a number of features in courseware which are behinds flags which may be disabled in the default config but are enabled on edx.org and those features may require a valid `request.user`. Is the anonymous access to courseware enabled on any course on edx.org that this set of changes can be tested with (on stage)?

symbolist commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -257,7 +257,18 @@ def student_view(self, context):
+        prereq_meta_info = {}
+
+        if self._required_prereq():
+            prereq_met, prereq_meta_info = self._compute_is_prereq_met(True)
```
@ormsbee I had to put this in during the rebase to match recent changes to the `student_view`: https://github.com/edx/edx-platform/commit/6fbd86108ce07484ce95dcc0df5ee836c40f8ccd#diff-14daf51ec8decc308953f7a4e7e0a03f

symbolist commented:
@ormsbee Thanks for the review! I'll resolve the conflicts.

symbolist commented on openedx/features/course_experience/tests/views/test_course_home.py:
```diff
@@ -239,37 +241,76 @@ def tearDown(self):
+        self.assertContains(
+            response, 'To see course content', count=(1 if is_anonymous else 0)
+        )
+        self.assertContains(response, '<div class="user-messages"', count=(1 if expected_enroll_message else 0))
```
Fixed conflict on this line because of https://github.com/edx/edx-platform/commit/67e9ae55817dee5d8b04f132515cd2bc55daaf42#diff-a9b73fc05b659b3ec3afab5a6f47b628R269

symbolist commented:
@ormsbee Okay, conflict resolved!

edx-webhook commented:
@symbolist 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/19284&entry.1484655318=https://github.com/symbolist&entry.752974735=2018-11-19+14:15&entry.1917517419=2018-12-10+16:03&entry.2133058324=Yes) so we can improve your experience in the future.
